### PR TITLE
feat(*): xds generator for MeshTrafficPermission

### PR DESCRIPTION
**Pull request by lobkovilya**
### Checklist prior to review

- [X] Link to docs PR or issue -- N/A
- [X] Link to UI issue or PR -- N/A
- [X] Is the [issue worked on linked][1]? -- https://github.com/kumahq/kuma/issues/4992
- [X] The PR does not hardcode values that might break projects that depend on kuma (e.g. "kumahq" as a image registry) --
- [X]  The PR will work for both Linux and Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [X] Unit Tests --
- [X] E2E Tests --
- [X] Manual Universal Tests --
- [X] Manual Kubernetes Tests --
- [X] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? -- No
- [X] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? -- No
- [X] Do you need to explicitly set a [`> Changelog:` entry here](../blob/master/CONTRIBUTING.md#submitting-a-patch)? No

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

**Comment by lahabana** [comment]
@lobkovilya can you rebase this?

**Comment by lobkovilya** [comment]
@lahabana I merged `master` yesterday, I think it didn't change since then

**Comment by lahabana** [comment]
Yes sorry noticed after...

File: pkg/xds/envoy/listeners/v3/rbac_configurer_test.go [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+			stats: "deny_all_prefix",
+			mesh:  "deny_all_mesh",
+			rules: []*core_xds.Rule{
+				{
+					Subset: []core_xds.Tag{},
+					Conf: &v1alpha1.MeshTrafficPermission_Conf{
+						Action: "DENY",
+					},
+				},
+			},
```
Comment (lahabana) [inline pkg/xds/envoy/listeners/v3/rbac_configurer_test.go:80]: Is there no way where there's no Rule at all?

File: pkg/xds/envoy/listeners/v3/rbac_configurer.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+			Principals: principals,
+		}
+	}
+
+	var shadowRules *rbac_config.RBAC
+	if shadowPrincipals := c.createPrincipals(true); len(shadowPrincipals) != 0 {
+		shadowRules = &rbac_config.RBAC{
+			Action: rbac_config.RBAC_ALLOW,
+			Policies: map[string]*rbac_config.Policy{
+				"ShadowMeshTrafficPermission": {
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/xds/envoy/listeners/v3/rbac_configurer.go:65]: I can rename if needed, I was just following the MADR https://github.com/kumahq/kuma/blob/master/docs/madr/decisions/007-mesh-traffic-permission.md?plain=1#L476

[the remark this answers is not in the document]
Comment (lahabana) [reply pkg/xds/envoy/listeners/v3/rbac_configurer.go:65]: I feel like `MeshTrafficPermission` is better suited here WDYT? (Update the MADR too if you end up changing it)

File: pkg/xds/envoy/listeners/v3/rbac_configurer.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+			Identifier: &rbac_config.Principal_AndIds{
+				AndIds: &rbac_config.Principal_Set{
+					Ids: principals,
+				},
+			},
+		}
+	}
+}
+
+func (c *RBACConfigurer) not(p *rbac_config.Principal) *rbac_config.Principal {
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/xds/envoy/listeners/v3/rbac_configurer.go:178]: `v3` package is pretty big, it contains a lot of configures, I think it better to scope short names like `not` within the RBACConfigurer type. We do the same for `generator` package, for example, method [generateLDS](https://github.com/kumahq/kuma/blob/master/pkg/xds/generator/outbound_proxy_generator.go#L95) could've been just a function

File: pkg/xds/envoy/listeners/v3/rbac_configurer.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/xds/envoy/listeners/v3/rbac_configurer.go:65]: Yeah, agreed, if it's already set as `shadowRules` no need to mention "shadow" twice

File: pkg/xds/envoy/listeners/v3/rbac_configurer.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+		}
+	}
+
+	rbacMarshalled, err := util_proto.MarshalAnyDeterministic(&rbac.RBAC{
+		// we include dot to change "inbound:127.0.0.1:21011rbac.allowed" metric to "inbound:127.0.0.1:21011.rbac.allowed"
+		StatPrefix: fmt.Sprintf("%s.", util_xds.SanitizeMetric(c.StatsName)),
+		Rules: &rbac_config.RBAC{
+			Action:   rbac_config.RBAC_ALLOW,
+			Policies: policies,
+		},
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/xds/envoy/listeners/v3/rbac_configurer.go]: There is a difference that `ShadowRules` can be `nil`, but `Rules` are always set (even if there are no principals). I tried to make it more symmetrical and added comments

File: pkg/xds/envoy/listeners/v3/rbac_configurer.go [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+		principals = append(principals, c.principalFromSubset(ss, c.Mesh))
+	}
+
+	return principals
+}
+
+func (c *RBACConfigurer) shouldGenerateRuleForAction(
+	action policies_api.MeshTrafficPermission_Conf_Action,
+	shadow bool,
+) bool {
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/xds/envoy/listeners/v3/rbac_configurer.go:126]: made it a function

File: pkg/xds/envoy/listeners/v3/rbac_configurer_test.go (hunk 1 above)
Comment (lobkovilya) [reply pkg/xds/envoy/listeners/v3/rbac_configurer_test.go:80]: Sorry, I'm not sure I'm following the question

File: pkg/xds/envoy/listeners/v3/rbac_configurer.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+	"fmt"
+
+	envoy_listener "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
+	rbac_config "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v3"
+	rbac "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/rbac/v3"
+	matcherv3 "github.com/envoyproxy/go-control-plane/envoy/type/matcher/v3"
+
+	mesh_proto "github.com/kumahq/kuma/api/mesh/v1alpha1"
+	core_xds "github.com/kumahq/kuma/pkg/core/xds"
+	policies_api "github.com/kumahq/kuma/pkg/plugins/policies/meshtrafficpermission/api/v1alpha1"
```
Comment (jakubdyszkiewicz) [inline pkg/xds/envoy/listeners/v3/rbac_configurer.go:13]: The fact that it has a dependency on `policies_api` is a bit weird. It makes it usable only with MeshTrafficPermission.

I'm ok with merging it as is, but as we go further with the following policies we need to think if this is expected or configurer should be placed in the plugin itself.

File: test/e2e_env/universal/meshtrafficpermission/meshtrafficpermission.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+          team: client-owners
+      default:
+        action: ALLOW
+`
+		err := YamlUniversal(yaml)(env.Cluster)
+		Expect(err).ToNot(HaveOccurred())
+
+		trafficAllowed()
+	})
+}
```
Comment (jakubdyszkiewicz) [inline test/e2e_env/universal/meshtrafficpermission/meshtrafficpermission.go:110]: I'd like to also see Kubernetes + Multizone test. Can be a separate PR.

File: pkg/xds/envoy/listeners/v3/rbac_configurer.go (hunk 6 above)
Comment (lahabana) [reply pkg/xds/envoy/listeners/v3/rbac_configurer.go:13]: That's a good point. @michaelbeaumont ended up putting the configurer in the plugins package. Maybe it's a cleaner approach?

File: docs/madr/decisions/007-mesh-traffic-permission.md [hunk 8]
```diff
@@ -473,7 +473,7 @@ List of principals generated based on the list from step 3. Each principal is ma
 ```yaml
 action: ALLOW
 policies:
-  "ShadowMeshTrafficPermission":
+  "MeshTrafficPermission":
```
Comment (lahabana) [inline docs/madr/decisions/007-mesh-traffic-permission.md:476]: 👍

File: pkg/xds/envoy/listeners/v3/rbac_configurer_test.go (hunk 1 above)
Comment (lahabana) [reply pkg/xds/envoy/listeners/v3/rbac_configurer_test.go:80]: Should there be a testCase:

```go
testCase{
  stats: "no",
  mesh: "nothing",
  rules: []*core_xds.Rule{Subset: []core_xds.Tag{}},
```

Or something like this?

File: pkg/xds/envoy/listeners/v3/rbac_configurer.go (hunk 6 above)
Comment (michaelbeaumont) [reply pkg/xds/envoy/listeners/v3/rbac_configurer.go:13]: If it's private to the package it also avoids an IMO unnecessary separate set of tests and we can solely test `Apply`.

File: pkg/plugins/policies/meshtrafficpermission/plugin/v1alpha1/plugin.go [hunk 9]
```diff
@@ -25,6 +30,49 @@ func (p plugin) MatchedPolicies(dataplane *core_mesh.DataplaneResource, resource
...
+			"mesh", ctx.Mesh.Resource.GetMeta().GetName())
+		return nil
+	}
+
+	mtp, ok := proxy.Policies.Dynamic[api.MeshTrafficPermissionType]
+	if !ok {
+		return nil
+	}
+
+	for _, res := range rs.Resources(envoy_resource.ListenerType) {
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/plugins/policies/meshtrafficpermission/plugin/v1alpha1/plugin.go:50]: Added a unit test for `Apply`

File: pkg/xds/envoy/listeners/v3/rbac_configurer_test.go (hunk 1 above)
Comment (lobkovilya) [reply pkg/xds/envoy/listeners/v3/rbac_configurer_test.go:80]: oh I get it, added a testCase `no rules`

File: pkg/plugins/policies/meshtrafficpermission/plugin/v1alpha1/plugin.go [hunk 10]
```diff
@@ -25,6 +30,49 @@ func (p plugin) MatchedPolicies(dataplane *core_mesh.DataplaneResource, resource
...
+	if proxy.Dataplane == nil {
+		// MeshTrafficPermission policy is applied only on DPP
+		return nil
+	}
+	if !ctx.Mesh.Resource.MTLSEnabled() {
+		log.V(1).Info("skip applying MeshTrafficPermission, MTLS is disabled",
+			"proxyName", proxy.Dataplane.GetMeta().GetName(),
+			"mesh", ctx.Mesh.Resource.GetMeta().GetName())
+		return nil
+	}
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/plugins/policies/meshtrafficpermission/plugin/v1alpha1/plugin.go:48]: makes sense, done

File: pkg/xds/envoy/listeners/v3/rbac_configurer.go (hunk 6 above)
Comment (lobkovilya) [reply pkg/xds/envoy/listeners/v3/rbac_configurer.go:13]: I agree to have this under `plugins/policies/meshtrafficpermission`, but I think it should be placed out of the `v1alpha1` package. Because Configures have their own versioning which depends on Envoy,  it's `v3` in our case. So it should probably look like:

policies/meshtrafficpermission:
* api
* k8s
* plugin
  * v1alpha1
* xds
  * v3

File: pkg/xds/envoy/listeners/v3/rbac_configurer.go [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+		// we include dot to change "inbound:127.0.0.1:21011rbac.allowed" metric to "inbound:127.0.0.1:21011.rbac.allowed"
+		StatPrefix:  fmt.Sprintf("%s.", util_xds.SanitizeMetric(c.StatsName)),
+		Rules:       rules,
+		ShadowRules: shadowRules,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &envoy_listener.Filter{
+		// todo(lobkovilya): use 'envoy.filters.http.rbac' for HTTP traffic to have proper stats
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/xds/envoy/listeners/v3/rbac_configurer.go]: https://github.com/kumahq/kuma/issues/5045

File: pkg/xds/envoy/listeners/v3/rbac_configurer.go (hunk 6 above)
Comment (lobkovilya) [reply pkg/xds/envoy/listeners/v3/rbac_configurer.go]: moved under `xds/v3`

File: mk/generate.mk [hunk 12]
```diff
@@ -45,7 +45,7 @@ cleanup/crds:
 
 # deletes all files in policy directory except *.proto and validator.go
 cleanup/policy/%:
-	$(shell find $(POLICIES_DIR)/$* \( -name '*.pb.go' -o -name '*.yaml' -o -name 'zz_generated.*'  \) -type f -delete)
+	$(shell find $(POLICIES_DIR)/$* \( -name '*.pb.go' -o -name '*.yaml' -o -name 'zz_generated.*'  \) -not -path '*/testdata/*' -type f -delete)
```
Comment (lahabana) [inline mk/generate.mk:48]: Might be better to include some `generated` name format in the generated files to make this better

File: pkg/plugins/policies/meshtrafficpermission/xds/v3/rbac_configurer_test.go [hunk 13]
```diff
@@ -0,0 +1,256 @@
+package v3_test
```
Comment (lahabana) [inline pkg/plugins/policies/meshtrafficpermission/xds/v3/rbac_configurer_test.go:1]: maybe get rid of v3 we've deprecated v2 and Matt Klein himself said that there'll likely never be a v4

File: test/e2e_env/universal/meshtrafficpermission/meshtrafficpermission.go (hunk 7 above)
Comment (lobkovilya) [reply test/e2e_env/universal/meshtrafficpermission/meshtrafficpermission.go:110]: https://github.com/kumahq/kuma/issues/5049

File: pkg/plugins/policies/meshtrafficpermission/xds/v3/rbac_configurer_test.go (hunk 13 above)
Comment (lobkovilya) [reply pkg/plugins/policies/meshtrafficpermission/xds/v3/rbac_configurer_test.go]: ok, I don't mind having just `meshtrafficpermission/xds`

File: pkg/plugins/policies/meshtrafficpermission/plugin/v1alpha1/plugin.go [hunk 14]
```diff
@@ -25,6 +30,50 @@ func (p plugin) MatchedPolicies(dataplane *core_mesh.DataplaneResource, resource
 }
 
 func (p plugin) Apply(rs *core_xds.ResourceSet, ctx xds_context.Context, proxy *core_xds.Proxy) error {
-	log.Info("apply is not implemented")
+	if proxy.Dataplane == nil {
+		// MeshTrafficPermission policy is applied only on DPP
```
[the remark this answers is not in the document]
Comment (lobkovilya) [reply pkg/plugins/policies/meshtrafficpermission/plugin/v1alpha1/plugin.go:34]: done

File: pkg/xds/envoy/listeners/v3/rbac_configurer.go (hunk 6 above)
Comment (lobkovilya) [reply pkg/xds/envoy/listeners/v3/rbac_configurer.go]: upd: moved under `xds`

**Review by jakubdyszkiewicz — APPROVED** [review]
